keps/3258-delayed-admission-check-retries/README.md

```diff
@@ -0,0 +1,545 @@
+
+```go
+type AdmissionCheckState struct {
+    // Existing fields remain unchanged...
+
+    // RetryAfter indicates when this check should be retried.
+    // When State=Retry, admission check controllers should set this
+    // to implement delays between retry attempts.
+    //
+    // If nil when State=Retry, Kueue will retry immediately.
+    // If set, Kueue will not re-evaluate until this time has passed.
+    //
+    // Example: To retry in 5 minutes:
+    //   retryAfter = &metav1.Time{Time: time.Now().Add(5*time.Minute)}
+    //
+    // +optional
+    RetryAfter *metav1.Time `json:"retryAfter,omitempty"`
+
+    // RetryCount tracks retry attempts for this admission check.
+    // Kueue automatically increments this when transitioning from
+    // Retry→Pending. Reset to 0 when check succeeds or workload
+    // is admitted.
+    //
+    // Controllers can use this for retry strategies:
+    // - Exponential backoff: delay = baseDelay * (2^retryCount)
+    // - Linear backoff: delay = baseDelay * retryCount
+    // - Max retries: if retryCount > limit, set State=Rejected
+    //
+    // +optional
+    RetryCount *int32 `json:"retryCount,omitempty"`
+}
+```
+
+#### Example: Workload Status with Multiple Admission Checks
+
+```yaml
+apiVersion: kueue.x-k8s.io/v1beta1
+kind: Workload
+metadata:
+  name: ml-training-job
+  namespace: research
+status:
+  conditions:
+    - type: QuotaReserved
+      status: "True"
+      lastTransitionTime: "2024-02-06T10:00:00Z"
+
+  admissionChecks:
+    # Check 1: Budget check with daily reset
+    - name: budget-check
+      state: Retry
+      lastTransitionTime: "2024-02-06T10:30:00Z"
+      message: "Daily budget exhausted. Will retry at midnight."
+      retryAfter: "2024-02-07T00:00:00Z" # Retry at midnight
+      retryCount: 1
+
+    # Check 2: GPU availability with exponential backoff
+    - name: gpu-availability
+      state: Retry
+      lastTransitionTime: "2024-02-06T10:30:00Z"
+      message: "No A100 GPUs available. Retrying in 8m (attempt 4)"
+      retryAfter: "2024-02-06T10:38:00Z" # 8 minutes from now
+      retryCount: 4
+
+    # Check 3: License server check - immediate retry
+    - name: license-check
+      state: Retry
+      lastTransitionTime: "2024-02-06T10:30:00Z"
+      message: "License server temporarily unavailable"
+      # No retryAfter set - will retry immediately
+# In this example, Kueue will wait until 00:00:00Z (the longest delay)
+# before re-evaluating all admission checks
+```
+
+### Controller Changes
+
+1. **Workload Controller** (`pkg/controller/core/workload_controller.go`):
+
+   - When processing admission checks in `Retry` state, calculate the maximum `retryAfter` across all checks
+   - Set workload requeue time based on the calculated delay
+   - Increment `retryCount` when transitioning back to `Pending`
+   - Remove the existing `requeueState` logic and migrate to per-admission-check retry handling
```

Comment:
Please note that the `requeueState` field in Workload's status in not used exclusively by Admission Checks but also WaitForPodsReady feature. This means while migrating AdmissionCheck related logic to per-admission-check retry handling, we should keep the remaining part related to WaitForPodsReady

Reply:
That's a good point and I wasn't aware of that. Thank you.
I'll take it into account in the updated version.